ISPN-8543 Pluggable global configuration persistence

tristantarrant opened the pull request:
https://issues.jboss.org/browse/ISPN-8543

Note: this doesn't actually implement any other configuration persistence backends yet. Also it does not expose the configuration to the XML parsers.

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/CacheState.java:
```diff
@@ -10,25 +10,32 @@
       this.flags = flags.clone();
    }
 
+   public String getTemplate() {
```
where is the template used?

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/EmbeddedLocalConfigurationManager.java:
```diff
@@ -0,0 +1,167 @@
+ * @since 9.2
+ */
+
+public class EmbeddedLocalConfigurationManager implements LocalConfigurationManager {
```
for some reason, I was expecting 2 implementations. 
one with persistence supported, and another without it...

tristantarrant replied:
I didn't feel like a simple ephemeral implementation was necessary...

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/EmbeddedLocalConfigurationManager.java:
```diff
@@ -0,0 +1,167 @@
+      }
+      if (flags.contains(CacheContainerAdmin.AdminFlag.PERSISTENT)) {
+         persistentCaches.add(name);
+         storeAll();
```
shouldn't the store be invoked when `add()` is successful?

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/EmbeddedLocalConfigurationManager.java:
```diff
@@ -0,0 +1,167 @@
+   public void createCache(String name, Configuration configuration, EnumSet<CacheContainerAdmin.AdminFlag> flags) {
+      Configuration existing = cacheManager.getCacheConfiguration(name);
+      if (existing == null) {
+         cacheManager.defineConfiguration(name, configuration);
```
should check the returned configuration is the same as `configuration` (or null)

tristantarrant replied:
Not quite sure I understand what you mean here. I already handle the case where the "new" config and the existing ones are equals, in the the second else clause by printing a debug log.

pruivo replied:
forget :)

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/EmbeddedLocalConfigurationManager.java:
```diff
@@ -0,0 +1,167 @@
+      log.debugf("Remove cache %s", name);
+      if (persistentCaches.contains(name)) {
+         persistentCaches.remove(name);
+         storeAll();
```
can be invoked if the remove is successful.

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/EmbeddedLocalConfigurationManager.java:
```diff
@@ -0,0 +1,167 @@
+         FileLock lock = null;
+         try (FileOutputStream lockFile = new FileOutputStream(getPersistentFileLock())) {
+            lock = lockFile.getChannel().lock();
+            temp.renameTo(getPersistentFile());
```
IMO, it should "warn" if the rename (delete below) fails.

tristantarrant replied:
I'd prefer to throw an exception.

pruivo replied:
are you throw the exception in this PR or another? :)

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java:
```diff
@@ -41,17 +32,20 @@
+
+   public static final String CACHE_SCOPE = "cache";
+
    private GlobalConfiguration globalConfiguration;
```
unused

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java:
```diff
@@ -41,17 +32,20 @@
    public void inject(GlobalConfiguration globalConfiguration, EmbeddedCacheManager cacheManager) {
       this.globalConfiguration = globalConfiguration;
       this.cacheManager = cacheManager;
+      this.localConfigurationManager = globalConfiguration.globalState().localConfigurationManager();
```
don't you want to create a factory for it? so it can be injected :)

pruivo commented on core/src/main/java/org/infinispan/globalstate/LocalConfigurationManager.java:
```diff
@@ -0,0 +1,51 @@
+   /**
+    * Loads all persisted cache configurations and instantiates them
+    */
+   void loadAll();
```
IMO, it should return a `Map<String, Configuration>` to avoid the cycli dependency `LocalConfigurationManager` <-> `GlobalConfigurationManager`

tristantarrant replied:
Right, much better that way

pruivo commented on core/src/main/java/org/infinispan/globalstate/LocalConfigurationManager.java:
```diff
@@ -0,0 +1,51 @@
+    * @param embeddedCacheManager
+    * @param globalConfigurationManager
+    */
+   void initialize(EmbeddedCacheManager embeddedCacheManager, GlobalConfigurationManager globalConfigurationManager);
```
I'm not sure if it is needed since you can use `@Inject` annotation to set dependencies...

tristantarrant replied:
I'm not treating it as a component in the component registry, and explicitly initializing it in GlobalConfigurationManagerImpl.postStart()

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java:
```diff
@@ -138,75 +99,38 @@ public Configuration createCache(String cacheName, String template, EnumSet<Cach
-      } catch (Exception e) {
-         throw log.errorPersistingGlobalConfiguration(e);
-      }
+   void removeCache(String name, CacheState state) {
```
the `state` is not needed.

pruivo replied:
name suggestion: `removeCacheLocally()` (or keep the old name)

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/GlobalConfigurationManagerImpl.java:
```diff
@@ -138,75 +99,38 @@ public Configuration createCache(String cacheName, String template, EnumSet<Cach
+      }
+   }
+
+   void createCache(String name, CacheState state) {
```
name suggestion: `createCacheLocally()` (or keep the old name)

pruivo commented on core/src/main/java/org/infinispan/configuration/global/GlobalStateConfiguration.java:
```diff
@@ -20,21 +22,26 @@
             .initializer(() -> SecurityActions.getSystemProperty("java.io.tmpdir")).immutable().build();
+   public static final AttributeDefinition<LocalConfigurationManager> LOCAL_CONFIGURATION_MANAGER = AttributeDefinition
+         .builder("localConfigurationManager", null, LocalConfigurationManager.class)
+         .initializer(() -> new EmbeddedLocalConfigurationManager()).immutable().build();
```
you can use method reference `EmbeddedLocalConfigurationManager::new`
althoug, using the class name doesn't sound a bad idea either...

tristantarrant commented:
@pruivo I've addressed the changes which you requested (unless otherwise noted).
I also had to introduce some additional changes in here:
- renamed the PERSISTENT flag to PERMANENT to avoid confusion with cache persistence (this is configuration persistence)
- Fix a logic problem in GlobalComponentRegistry PostStart handling (ISPN-8625)
- Unwrap ExecutionExceptions in the Hot Rod server to expose the real cause to clients (ISPN-8626)
- Add a getOrCreateCache() method to CacheContainerAdmin (ISPN-8627): I wanted to separate the commit but I've reworked some of the logic in the GlobalConfigurationManager code around it

pruivo commented on core/src/test/java/org/infinispan/manager/CacheManagerAdminTest.java:
```diff
@@ -30,22 +32,37 @@ public void testClusterCacheTest() {
       checkConsistencyAcrossCluster("a", configuration);
 
-      manager(1).administration().removeCache("a");
+      try {
```
suggestion:
```
Exceptions.expectException(CacheConfigurationException.class,
            "Should not succeed in creating a cache with a non-existing template",
            () -> manager(0).administration().createCache("b", "nonExistingTemplate"));
```
also, there is a typo in `fail()` message (succeed)

pruivo commented on core/src/test/java/org/infinispan/manager/CacheManagerAdminTest.java:
```diff
@@ -30,22 +32,37 @@ public void testClusterCacheTest() {
 
-      checkCacheExistenceAcrossCluster("a", false);
+      // attempt to create an existing cache
+      try {
```
suggestion:
```
Exceptions.expectException(CacheConfigurationException.class,
            "Should not have allowed creation of existing cache",
            () -> manager(0).administration().createCache("a", configuration));
```

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/CacheState.java:
```diff
@@ -41,15 +48,17 @@ public String getConfiguration() {
+         String configuration = MarshallUtil.unmarshallString(input);
          EnumSet<CacheContainerAdmin.AdminFlag> flags = (EnumSet<CacheContainerAdmin.AdminFlag>) input.readObject();
-         return new CacheState(configuration, flags == null ? EnumSet.noneOf(CacheContainerAdmin.AdminFlag.class) : flags);
+         return new CacheState(template, configuration, flags == null ? EnumSet.noneOf(CacheContainerAdmin.AdminFlag.class) : flags);
```
I don't think `flags` can be `null`.

pruivo commented on core/src/main/java/org/infinispan/manager/EmbeddedCacheManagerAdmin.java:
```diff
@@ -26,4 +48,18 @@
     * @return the cache
     */
    <K, V> Cache<K, V> createCache(String name, Configuration configuration);
```
I guess this method will also throw a `CacheException` if the cache already exists (missing in javadoc)

pruivo commented on core/src/main/java/org/infinispan/globalstate/impl/EmbeddedLocalConfigurationManager.java:
```diff
@@ -0,0 +1,170 @@
+   private EmbeddedCacheManager cacheManager;
+   private ParserRegistry parserRegistry;
+   private GlobalConfiguration globalConfiguration;
+   private GlobalConfigurationManager globalConfigurationManager;
```
unused

tristantarrant commented:
@pruivo addressed all issues

pruivo commented on core/src/test/java/org/infinispan/manager/CacheManagerAdminTest.java:
```diff
@@ -4,11 +4,14 @@
 import static org.testng.AssertJUnit.assertFalse;
 import static org.testng.AssertJUnit.assertNotNull;
 import static org.testng.AssertJUnit.assertTrue;
+import static org.testng.AssertJUnit.fail;
```
unused import

pruivo commented on core/src/test/java/org/infinispan/configuration/serializer/ConfigurationSerializerTest.java:
```diff
@@ -9,7 +9,7 @@
    public Object[][] configurationFiles() {
       return new Object[][] {
-            {"configs/unified/7.0.xml"},
+            {"configs/unified/7.0.S"},
```
oops!?

pruivo commented on core/src/main/java/org/infinispan/globalstate/GlobalConfigurationManager.java:
```diff
@@ -24,20 +24,38 @@
     * @param flags the flags to apply
     */
-   Configuration createCache(String cacheName, Configuration configuration, EnumSet<CacheContainerAdmin.AdminFlag> flags);
+   Configuration createCache(String cacheName, Configuration configuration, EnumSet<AdminFlag> flags);
```
to think (in another PR): the return values aren't used. probably they should return void.

tristantarrant replied:
Yes, I'm trying to remember why I did this, and I think it's connected to the persistence of configuration into the server model, but I need to fetch that from the back of my brain.

pruivo replied:
no worries :+1:

pruivo commented:
let's wait for CI before pushing the button :)

tristantarrant commented:
I had to revert the CacheState externalizer reader check for nullness on the EnumSet, as the EnumSet externalizer actually returns null for empty sets...

pruivo commented on commons/src/main/java/org/infinispan/commons/configuration/attributes/Attribute.java:
```diff
@@ -198,8 +198,15 @@ public boolean matches(Attribute<?> other) {
+      if (this.definition.isGlobal()) {
+         if (Matchable.class.isAssignableFrom(this.definition.getType())) {
+            return ((Matchable)value).matches(other.value);
+         } else if (value != null) {
```
suggestion
```
} else {
return Objects.equals(value, other.value);
} 
```

pruivo commented on server/core/src/main/java/org/infinispan/server/core/admin/embeddedserver/CacheGetOrCreateTask.java:
```diff
@@ -0,0 +1,56 @@
+ * </ul>
+ *
+ * @author Tristan Tarrant
+ * @since 9.1
```
I guess I miss this: `@since 9.2`

pruivo commented on core/src/main/java/org/infinispan/configuration/cache/ClusteringConfiguration.java:
```diff
@@ -136,6 +136,41 @@ public AttributeSet attributes() {
+            return false;
+      } else if (!attributes.matches(other.attributes))
+         return false;
+      if (hashConfiguration == null) {
```
is it possible to the `*Configuration` to be `null`?

tristantarrant replied:
I copied equals and s/equals/matches/ :)

pruivo replied:
suggestion:
```
return hashConfiguration.matches(other.hashConfiguration) &&
          l1Configuration.matches(other.l1Configuration) &&
          ...
```

pruivo replied:
ok

tristantarrant commented:
Done @pruivo

pruivo commented:
failures not related. integrated! thanks @tristantarrant
